src/plugins/backend_odbc/backend_odbc_get.c

```diff
@@ -0,0 +1,842 @@
+ * @param[out] errorKey Used to store errors and warnings
+ *
+ * @returns The KeySet with the data retrieved from the SQL SELECT query.
+ * 	Make sure to ksDel() the returned KeySet.
+ * @retval NULL if no data was fetched or an error occurred (see @p errorKey for details)
+ *
+ * @see executeSqlStatement() for executing a prepared SQL statement
+ */
+static KeySet * fetchResults (SQLHSTMT sqlStmt, struct columnData * buffers, Key * errorKey)
+{
+	SQLRETURN ret;
+	KeySet * ksResult = NULL;
+
+	/* Needed for detecting if the current row contains a new key or just a further metakey for a previous key (outer join) */
+	char * prevKeyName = NULL;
+	Key * curKey = NULL;
+	Key * prevKey = NULL;
+
+	if (!sqlStmt || !buffers)
+	{
+		if (errorKey)
+		{
+			ELEKTRA_SET_INTERFACE_ERROR (
+				errorKey, "A NULL pointer was given as an argument to fetchResults() for 'sqlStmt' or 'columnData'");
+		}
+		return NULL;
+	}
+
+	while ((ret = SQLFetch (sqlStmt)) != SQL_NO_DATA)
+	{
+
+		if (!SQL_SUCCEEDED (ret))
+		{
+			ELEKTRA_SET_ODBC_ERROR (SQL_HANDLE_STMT, sqlStmt, errorKey);
+			elektraFree (prevKeyName);
+
+			if (ksResult)
+			{
+				ksDel (ksResult);
+			}
+
+			SQLFreeHandle (SQL_HANDLE_STMT, sqlStmt);
+			return NULL;
+		}
+
+		ELEKTRA_ASSERT (buffers->bufferKeyName[0] && buffers->nameLenInd != SQL_NULL_DATA,
+				"The ODBC-backend retrieved an empty- or null-string for the key-name. This is not allowed! Please check "
+				"you data source.");
+
+		char * longKeyName = NULL;
+		char * longKeyString = NULL;
+		char * longMetaKeyName = NULL;
+		char * longMetaKeyString = NULL;
+		bool isFurtherMetaKey = false;
+
+
+		if (ret == SQL_SUCCESS_WITH_INFO)
+		{
+			/* Check if string was truncated (buffer too small) */
+			SQLCHAR sqlState[SQL_SQLSTATE_SIZE + 1];
+			SQLGetDiagField (SQL_HANDLE_STMT, sqlStmt, 1, SQL_DIAG_SQLSTATE, sqlState, SQL_SQLSTATE_SIZE + 1, 0);
+
+
+			/* SQLSTATE 01004 = Data truncated */
+			if (elektraStrCmp ((const char *) sqlState, "01004") == 0)
+			{
+				/* The buffer size constants include the byte for the \0, while the LenInd-variables don't! */
+
+				/* Check key-name column */
+				bool retGetLongData = true;
+				if (buffers->nameLenInd == SQL_NO_TOTAL)
+				{
+					/* No information about the total length is available --> The new buffer maybe must be
+					 * resized multiple times */
+					longKeyName = (char *) elektraMalloc (sizeof (char) * (KEYNAME_BUFFER_SIZE * 2));
+					retGetLongData =
+						getLongData (sqlStmt, 1, SQL_C_CHAR, &longKeyName, KEYNAME_BUFFER_SIZE * 2, errorKey);
+				}
+				else if (KEYNAME_BUFFER_SIZE <= buffers->nameLenInd)
+				{
+					longKeyName = (char *) elektraMalloc (sizeof (char) * (buffers->nameLenInd + 1));
+					retGetLongData =
+						getLongData (sqlStmt, 1, SQL_C_CHAR, &longKeyName, (buffers->nameLenInd + 1), errorKey);
+				}
+
+				/* Check if the keyname changed since the last iteration */
+				if (longKeyName && prevKeyName && elektraStrCmp (longKeyName, prevKeyName) == 0)
```

Comment:
That seems really inefficient. But I don't have a suggestions for a better solution (other than N+1 `SELECT`).